### PR TITLE
Add sip connect dial disconnect

### PR DESCRIPTION
**Pull request by kevingarabedian2**

### CODE REVIEW

File: signalwire-dotnet/Relay/CallingAPI.cs [hunk 1]
```diff
@@ -72,13 +72,13 @@ public SipCall NewSipCall(string to, string from, string fromName = null, string
...
+        public DialAction DialPhoneAsync(string to, string from, string region = null) { return NewPhoneCall(to, from, region).DialAsync(); }
 
-        public DialResult DialSip(string to, string from, string fromName = null, string codecs = null, JArray headers = null, int? maxDuration = null) { return NewSipCall(to, from, fromName,codecs, headers, maxDuration).Dial(); }
+        public DialResult DialSip(string to, string from, string fromName = null, string codecs = null, JArray headers = null, string region = null) { return NewSipCall(to, from, fromName,codecs, headers, region).Dial(); }
 
-        public DialAction DialSipAsync(string to, string from, string fromName = null, string codecs = null, JArray headers = null, int timeout = 30, int? maxDuration = null) { return NewSipCall(to, from, fromName, codecs, headers, maxDuration).DialAsync(); }
+        public DialAction DialSipAsync(string to, string from, string fromName = null, string codecs = null, JArray headers = null, string region = null) { return NewSipCall(to, from, fromName, codecs, headers, region).DialAsync(); }
```
Comment (Astaelan) [inline signalwire-dotnet/Relay/CallingAPI.cs:82]: Are these parameters actually removed? timeout and maxDuration were pretty important to solve calls that don't get hung up properly. Follow up to make sure these were removed on the system side of the call, I suspect they should still be there, @crienzo might be able to confirm.

Comment (kevingarabedian2) [reply signalwire-dotnet/Relay/CallingAPI.cs:81]: I just pushed the more recent code from this morning, Timeout exists and is in there, but MaxDuration is not in dial as a param per spec.

Comment (crienzo) [reply signalwire-dotnet/Relay/CallingAPI.cs:81]: Looks like we forgot to add it to the spec.

File: signalwire-dotnet/Relay/Calling/Call.cs [hunk 2]
```diff
@@ -1944,31 +1996,45 @@ protected override async Task<DialResult> InternalDialAsync()
...
 
                 // The use of await rethrows exceptions from the task
-                LL_BeginResult resultLLBegin = await taskLLBegin;
-                if (resultLLBegin.Code == "200")
+                LL_DialResult resultLLDial = await taskLLDial;
+                 Log(LogLevel.Debug, "did the task start....");
+                if (resultLLDial.Code == "200")
                 {
-                    Log(LogLevel.Debug, string.Format("Dial for call {0} waiting for completion events", ID));
+                    Log(LogLevel.Debug, string.Format("Dial kevinfor call {0} waiting for completion events", ID));
```
Comment (Astaelan) [inline signalwire-dotnet/Relay/Calling/Call.cs:2031]: Search the code for "kevin" I've seen references like this in a couple spots.

File: signalwire-dotnet/Relay/Calling/Call.cs [hunk 3]
```diff
@@ -1944,31 +1996,45 @@ protected override async Task<DialResult> InternalDialAsync()
 
             try
             {
-                Task<LL_BeginResult> taskLLBegin = mAPI.LL_BeginAsync(new LL_BeginParams()
```
Comment (Astaelan) [inline signalwire-dotnet/Relay/Calling/Call.cs:1947]: Just above here the events are being hooked for the old Begin based dialing. My understanding is that the new Dial is a near verbatim version of Connect. As per https://github.com/signalwire/relay-apis/blob/master/public/signalwire_calling.md#callingcalldial the possible states are "dialing", "answered" and "failed".
So these events need to be propagated similarly to how Connect does as follows:
1) Add OnCallingEvent_Dial to CallingAPI similar to OnCallingEvent_Connect and call it from CallingAPI.OnNotification, 2) Add DialHandler similar to ConnectHandler in Call and call it from OnCallingEvent_Dial, 3) Add OnDialX event variables to Call similar to OnConnectX events, and call them from DialHandler, 4) Assign the events above here in Dial to allow tcsCompletion to be set on Answered or Failed.

File: signalwire-dotnet/Relay/Calling/Call.cs [hunk 4]
```diff
@@ -1944,31 +1996,45 @@ protected override async Task<DialResult> InternalDialAsync()
...
+                if (resultLLDial.Code == "200")
                 {
-                    Log(LogLevel.Debug, string.Format("Dial for call {0} waiting for completion events", ID));
+                    Log(LogLevel.Debug, string.Format("Dial kevinfor call {0} waiting for completion events", ID));
 
                     resultDial.Successful = await tcsCompletion.Task;
 
                     Log(LogLevel.Debug, string.Format("Dial for call {0} {1}", ID, resultDial.Successful ? "successful" : "unsuccessful"));
+
+                    Log(LogLevel.Debug, "did the task complete....");
```
Comment (Astaelan) [inline signalwire-dotnet/Relay/Calling/Call.cs:2037]: Don't think we need this log line here, we have one right above.

File: signalwire-dotnet/Relay/Calling/Call.cs [hunk 5]
```diff
@@ -1944,31 +1996,45 @@ protected override async Task<DialResult> InternalDialAsync()
...
+                                },
+                    Region = Region,
+                    Tag = Tag
                 });
 
                 // The use of await rethrows exceptions from the task
-                LL_BeginResult resultLLBegin = await taskLLBegin;
-                if (resultLLBegin.Code == "200")
+                LL_DialResult resultLLDial = await taskLLDial;
+                 Log(LogLevel.Debug, "did the task start....");
```
Comment (Astaelan) [inline signalwire-dotnet/Relay/Calling/Call.cs:2028]: More left over log lines we don't need to keep.

**Comment by kevingarabedian2** [comment]
Thanks for your feedback Shane, I will make some changes and we can review again AM.

File: signalwire-dotnet/Relay/Calling/Call.cs [hunk 6]
```diff
@@ -2019,31 +2087,40 @@ protected override async Task<DialResult> InternalDialAsync()
 
             try
             {
-                Task<LL_BeginResult> taskLLBegin = mAPI.LL_BeginAsync(new LL_BeginParams()
```
Comment (Astaelan) [inline signalwire-dotnet/Relay/Calling/Call.cs:2022]: See above in SipCall for details about events and Dial implementation updates.

File: signalwire-dotnet/Relay/Calling/Call.cs [hunk 7]
```diff
@@ -1908,11 +1954,17 @@ private async Task<SendDigitsResult> InternalSendDigitsAsync(string controlID, s
         }
     }
 
-public sealed class SipCall : Call
+    public sealed class SipCall : Call
     {
         public string To { get; set; }
         public string From { get; set; }
-        public JObject Headers { get; set; }
+        public string FromName { get; set; }
```
Comment (Astaelan) [inline signalwire-dotnet/Relay/Calling/Call.cs:1961]: If Dial can now take multiple devices, are we expected to expose this to the end users? If so, a single set of these variables is not enough to initiate a call with multiple targets and will need to be packed into a `List<List<X>>` where X represents a type containing variables relative to each target of the Dial.

**Comment by emcgee** [comment]
@kevingarabedian2 Just as a note - for commit messages, we want to make sure they actually reflect what in the code has changed. Even in PRs that are going to be squashed they need to be informative but concise - feel free to expand in the additional section if you like. 

In this particular repo, take a look at https://github.com/signalwire/signalwire-dotnet/pull/95 or other closed pull requests for some good examples.

One easy way to do this is to plan out each next step of the development before even working on the code: 

- Take a look at what's necessary from the high level
- "Now I'll add the following method/function" or "I'll fix the following issue"
- Commit message should then reflect as closely as possible what was done
- If in any doubt, follow the guidelines at https://github.com/signalwire/meta/wiki/Git-Development-Guidelines

**Comment by kevingarabedian2** [comment]
Thanks for note Evan! I will make future commits more descriptive.  I have been pushing work more frequently and got into that bad habbit of using 'save'.  I totally see the benefit of quality descriptions and will make corrections.  Hope you have a wonderful weekend!

**Comment by Astaelan** [comment]
> @kevingarabedian2 Just as a note - for commit messages, we want to make sure they actually reflect what in the code has changed. Even in PRs that are going to be squashed they need to be informative but concise - feel free to expand in the additional section if you like. `save` isn't going to be useful.
> 
> In this particular repo, take a look at #95 or other closed pull requests for some good examples.
> 
> One easy way to do this is to plan out each next step of the development before even working on the code:
> 
> * Take a look at what's necessary from the high level
> * "Now I'll add the following method/function" or "I'll fix the following issue"
> * `<insert coding>`
> * Commit message should then reflect as closely as possible what was done
> * If in any doubt, follow the guidelines at https://github.com/signalwire/meta/wiki/Git-Development-Guidelines

This is particularly useful near the end when it comes time to doing a packaged release.  You can go back over the commit notes and squash down the useful bits for the CHANGELOG.md.

**Review by DonLoquacious — DISMISSED** [review]

**Review by DonLoquacious — APPROVED** [review]

**Pull request merged**